apex/transformer/parallel_state.py

```diff
@@ -80,6 +79,73 @@ def is_unitialized():
     """Useful for code segments that may be accessed with or without mpu initialization"""
     return _DATA_PARALLEL_GROUP is None
 
+def set_nccl_socket_envs():
+    assert os.getenv("NCCL_SOCKET_IFNAME") is not None, "NCCL_SOCKET_IFNAME was not set"
+    os.environ["NCCL_NET"] = "Socket"
+
+def set_nccl_ib_envs():
+    os.environ["NCCL_NET"] = "IB"
+
+def init_nccl_net(group, device="cuda"):
+    temp = torch.ones(1, device=device)
+    torch.distributed.all_reduce(temp, group=group)
+    torch.cuda.synchronize()
+
+def new_nccl_socket_group(ranks, backend):
+    set_nccl_socket_envs()
+    group = torch.distributed.new_group(ranks, backend=backend)
+    init_nccl_net(group=group)
+    return group
+
+def new_nccl_ib_group(ranks, backend):
+    set_nccl_ib_envs()
+    group = torch.distributed.new_group(ranks, backend=backend)
+    init_nccl_net(group=group)
+    return group
+
+def new_process_group(ranks, backend):
+    """
+    This function creates process groups.
+
+    In addition to simply creating the process groups, it initializes NCCL
+    for hybrid IB/Socket network like in the following diagram:
+
+      _______________________________
+      |          |       |          |
+     TCP        TCP     TCP        TCP
+      |          |       |          |
+    GPU:0--IB--GPU:1 | GPU:2--IB--GPU:3
+
+
+    If an environment variable NUM_GPUS_PER_IB_BLOCK is defined it looks up the ranks
+    and determines whether the set of ranks belong to the same computational block where
+    GPUs are interconnected via IB type of connection or not.
+    If all ranks are in the same block, the process group will use NCCL_NET=IB for
+    communication otherwise it will use NCCL_NET=Socket. 
+    
+    If NCCL_NET=Socket is ever to be used, the user must set NCCL_SOCKET_IFNAME.
+    Additionally, it is recommended to set NCCL_SOCKET_NTHREADS and
+    NCCL_NSOCKS_PERTHREAD before running the job.
+    See: https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/env.html
+    for more info
```

Comment:
I have one newbie question. What will happen if Socket is used without `NCCL_SOCKET_IFNAME` specified by a user?

Reply:
It is some sort of an undefined behavior. It may work if there is only one interface. However, if there are more than one interfaces (eg. `eth0`, `eth1`, `eth2` ...) it will hang or throw an error in best scenario.

Reply:
hmm, sounds like it (= Socket is used but the env var is missing) deserves RuntimeError if it could lead to unknown status